src/main/java/com/hubspot/jinjava/lib/tag/IfTag.java

```diff
@@ -58,46 +58,51 @@ public String interpret(TagNode tagNode, JinjavaInterpreter interpreter) {
       throw new TemplateSyntaxException(tagNode.getMaster().getImage(), "Tag 'if' expects expression", tagNode.getLineNumber(), tagNode.getStartPosition());
     }
 
+    LengthLimitingStringBuilder sb = new LengthLimitingStringBuilder(interpreter.getConfig().getMaxOutputSize());
+
     Iterator<Node> nodeIterator = tagNode.getChildren().iterator();
-    TagNode nextIfElseTagNode = tagNode;
 
-    while (nextIfElseTagNode != null && !evaluateIfElseTagNode(nextIfElseTagNode, interpreter)) {
-      nextIfElseTagNode = findNextIfElseTagNode(nodeIterator);
-    }
+    boolean parentValidationMode = interpreter.getContext().isValidationMode();
+
+    boolean execute = isPositiveIfElseNode(tagNode, interpreter);
+    boolean executedAnyBlock = false;
+
+    try {
 
-    LengthLimitingStringBuilder sb = new LengthLimitingStringBuilder(interpreter.getConfig().getMaxOutputSize());
-    if (nextIfElseTagNode != null) {
       while (nodeIterator.hasNext()) {
-        Node n = nodeIterator.next();
-        if (n.getName().equals(ElseIfTag.ELSEIF) || n.getName().equals(ElseTag.ELSE)) {
-          break;
+
+        executedAnyBlock = executedAnyBlock || execute;
+        if (interpreter.isValidationMode() && !parentValidationMode) {
+          interpreter.getContext().setValidationMode(!execute);
         }
-        sb.append(n.render(interpreter));
-      }
-    }
 
-    return sb.toString();
-  }
+        Node node = nodeIterator.next();
+        if (TagNode.class.isAssignableFrom(node.getClass())) {
+          TagNode tag = (TagNode) node;
+          if (tag.getName().equals(ElseIfTag.ELSEIF)) {
+            execute = isPositiveIfElseNode(tag, interpreter);
+            continue;
+          } else if (tag.getName().equals(ElseTag.ELSE)) {
+            execute = !executedAnyBlock;
+            continue;
+          }
+        }
 
-  private TagNode findNextIfElseTagNode(Iterator<Node> nodeIterator) {
-    while (nodeIterator.hasNext()) {
-      Node node = nodeIterator.next();
-      if (TagNode.class.isAssignableFrom(node.getClass())) {
-        TagNode tag = (TagNode) node;
-        if (tag.getName().equals(ElseIfTag.ELSEIF) || tag.getName().equals(ElseTag.ELSE)) {
-          return tag;
+        if (execute) {
+          sb.append(node.render(interpreter));
+        } else if (interpreter.getContext().isValidationMode()) {
+          node.render(interpreter);
         }
       }
+
+    } finally {
+      interpreter.getContext().setValidationMode(parentValidationMode);
     }
 
-    return null;
+    return sb.toString();
   }
 
-  protected boolean evaluateIfElseTagNode(TagNode tagNode, JinjavaInterpreter interpreter) {
-    if (tagNode.getName().equals(ElseTag.ELSE)) {
```

Comment:
Is this checking hurting?

Reply:
I don't think it's necessary anymore. `isPositiveIfElseNode` is only called once at the top of the `if` statement and again when an `elseif` is encoutered.